File: src/main/java/ru/arc/board/ItemIcon.java
package ru.arc.board;

import ru.arc.util.HeadUtil;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.SkullMeta;

import java.util.Objects;
import java.util.UUID;

@Getter @Setter
@NoArgsConstructor
@AllArgsConstructor
public class ItemIcon {

    Material material;
    UUID headUuid;
    int modelData = 0;

    public static ItemIcon of(UUID uuid){
        ItemIcon icon = new ItemIcon();
        icon.setMaterial(Material.PLAYER_HEAD);
        icon.setHeadUuid(uuid);
        return icon;
    }


    public static ItemIcon of(Material material, int modelData){
        ItemIcon icon = new ItemIcon();
        icon.setMaterial(material);
        icon.setModelData(modelData);
        return icon;
    }

    public static ItemIcon of(ItemStack stack){
        ItemIcon icon = new ItemIcon();
        icon.setMaterial(stack.getType());
        if(stack.getItemMeta().hasCustomModelData()) icon.setModelData(stack.getItemMeta().getCustomModelData());
        if(stack.getType() == Material.PLAYER_HEAD){
            System.out.println(stack.serialize());
            SkullMeta skullMeta = (SkullMeta)stack.getItemMeta();
            if(skullMeta.getOwningPlayer() != null) {
                icon.setHeadUuid(skullMeta.getOwningPlayer().getUniqueId());
            }
        }
        return icon;
    }

    public ItemStack stack(){
        if(material == Material.PLAYER_HEAD) return HeadUtil.getSkull(headUuid);
        ItemStack stack = new ItemStack(material);
        if(modelData != 0) {
            ItemMeta meta = stack.getItemMeta();
            meta.setCustomModelData(modelData);
            stack.setItemMeta(meta);
        }
        return stack;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ItemIcon icon)) return false;

        if (modelData != icon.modelData) return false;
        if (material != icon.material) return false;
        return Objects.equals(headUuid, icon.headUuid);
    }

    @Override
    public int hashCode() {
        int result = material != null ? material.hashCode() : 0;
        result = 31 * result + (headUuid != null ? headUuid.hashCode() : 0);
        result = 31 * result + modelData;
        return result;
    }
}

File: src/main/kotlin/ru/arc/board/ItemIcon.kt
package ru.arc.board

import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta
import ru.arc.util.HeadUtil
import ru.arc.util.customModelDataOrNull
import java.util.UUID

/**
 * Represents an item icon for a board entry — either a material with optional custom model data,
 * or a player head identified by UUID.
 */
data class ItemIcon(
    val material: Material,
    val headUuid: UUID?,
    val modelData: Int
) {
    /** No-args constructor for Jackson/Gson deserialization compatibility. */
    constructor() : this(Material.STONE, null, 0)

    fun stack(): ItemStack {
        if (material == Material.PLAYER_HEAD) {
            return headUuid?.let { HeadUtil.getSkull(it) } ?: ItemStack(Material.PLAYER_HEAD)
        }
        val stack = ItemStack(material)
        if (modelData != 0) {
            @Suppress("DEPRECATION")
            stack.editMeta { it.setCustomModelData(modelData) }
        }
        return stack
    }

    companion object {
        @JvmStatic
        fun of(uuid: UUID): ItemIcon = ItemIcon(material = Material.PLAYER_HEAD, headUuid = uuid, modelData = 0)

        @JvmStatic
        fun of(material: Material, modelData: Int): ItemIcon = ItemIcon(material = material, headUuid = null, modelData = modelData)

        @JvmStatic
        fun of(stack: ItemStack): ItemIcon {
            val meta = stack.itemMeta
            val customModelData = meta?.customModelDataOrNull ?: 0
            val headUuid = if (stack.type == Material.PLAYER_HEAD) {
                (meta as? SkullMeta)?.owningPlayer?.uniqueId
            } else null
            return ItemIcon(material = stack.type, headUuid = headUuid, modelData = customModelData)
        }
    }
}
